sources/rest_api/client.py

```diff
@@ -0,0 +1,109 @@
+from typing import Optional, Dict, Any, Generator, Literal
+import copy
+
+from requests.auth import AuthBase
+from requests import Session as BaseSession
+
+
+from dlt.common import logger
+from dlt.sources.helpers.requests import client
+
+from .paginators import BasePaginator, UnspecifiedPaginator
+from .detector import create_paginator
+
+from .utils import join_url
+
+class RESTClient:
+    """A generic REST client for making requests to an API.
+
+    Attributes:
+        base_url (str): The base URL of the API.
+        headers (Optional[Dict[str, str]]): Headers to include in all requests.
+        auth (Optional[AuthBase]): An authentication object to use for all requests.
+        paginator (Optional[BasePaginator]): A paginator object for handling API pagination.
+            Note that this object will be deepcopied for each request to ensure that the
+            paginator state is not shared between requests.
+    """
+
+    def __init__(
+        self,
+        base_url: str,
+        headers: Optional[Dict[str, str]] = None,
+        auth: Optional[AuthBase] = None,
+        paginator: Optional[BasePaginator] = None,
+        session: BaseSession = None
+    ) -> None:
+        self.base_url = base_url
+        self.headers = headers
+        self.auth = auth
+        self.session = session or client.session
+        self.paginator = paginator if paginator else UnspecifiedPaginator()
+
+    def make_request(self, path="", method="get", params=None, json=None):
+        if path.startswith("http"):
+            url = path
+        else:
+            url = join_url(self.base_url, path)
+
+        logger.info(
+            f"Making {method.upper()} request to {url} with params={params}, "
+            f"json={json}"
+        )
+
+        response = self.session.request(
+            method=method,
+            url=url,
+            headers=self.headers,
+            params=params if method.lower() == "get" else None,
+            json=json if method.lower() in ["post", "put"] else None,
+            auth=self.auth,
+        )
+        response.raise_for_status()
```

Comment:
instead call an abstract method that people can subclass. `response.raise_for_status()` should be default implementation